Implement ycm_quiet_messages options (See #2021)

dhleong opened the pull request:
# PR Prelude

- [x] I have read and understood YCM's [CONTRIBUTING][cont] document.
- [x] I have read and understood YCM's [CODE_OF_CONDUCT][code] document.
- [x] I have included tests for the changes in my PR. If not, I have included a
  rationale for why I haven't.
- [x] **I understand my PR may be closed if it becomes obvious I didn't
  actually perform all of these steps.**

# Why this change is necessary and useful

See issue #2021. This is a partial implementation based on the syntastic option referenced, supporting the `!` flag and filters of type `regex` and `level`. Also supports filetype specific filters using `ycm_<ft>_quiet_messages`, but I couldn't think of a great way to fall back to syntastic configs for this one.

In terms of usefulness: I've been playing with C# recently, which has a bunch of style warnings that I don't want to follow, and prefer to only have the gutter showing if there are actually errors, or warnings that I *do* want to follow.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/valloric/youcompleteme/2377)
<!-- Reviewable:end -->

Valloric commented:
Thanks for the PR! :)

I agree this is a feature we should have.

Travis runs are failing for this PR, so you'll need to look into that. 

I gave a light review; the code doesn't quite match the style we follow in the rest of the codebase. I suggest looking through other files to identify what the style looks like. That might be faster than us pointing out every style violation. :)

Also, this needs to change the README so that we have documentation for the new option.

---

Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[python/ycm/diagnostic_filter.py, line 1 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTkd6g70Aoq1XQS4g3s:-KTkd6g8tS1lCzJCXZCc:b-boinl4) ([raw file](https://github.com/valloric/youcompleteme/blob/8263773ccde05a323443fa14f1c55d99ae88d26b/python/ycm/diagnostic_filter.py#L1)):*
> ```Python
> # Copyright (C) 2013  Google Inc.
> ```

The copyright line is wrong. Update year and author is "YouCompleteMe contributors".

---

*[python/ycm/diagnostic_filter.py, line 41 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTkdNyA0frVaC5k1iCd:-KTkdNyA0frVaC5k1iCe:b69sgpp) ([raw file](https://github.com/valloric/youcompleteme/blob/8263773ccde05a323443fa14f1c55d99ae88d26b/python/ycm/diagnostic_filter.py#L41)):*
> ```Python
>       actual_filter_type = filter_type
> 
>       if filter_type[0] == '!':
> ```

Whitespace inside brackets.

---

*[python/ycm/diagnostic_filter.py, line 45 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTkdEqYZBHPtqGNPCAI:-KTkdEqYZBHPtqGNPCAJ:b-gak7ss) ([raw file](https://github.com/valloric/youcompleteme/blob/8263773ccde05a323443fa14f1c55d99ae88d26b/python/ycm/diagnostic_filter.py#L45)):*
> ```Python
>         filter_type = filter_type[1:]
>       compiler = FILTER_COMPILERS.get( filter_type )
>       
> ```

No trailing whitespace please. :)

---

*[python/ycm/tests/diagnostic_filter_tests.py, line 1 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTkdWNH7C-f99oTCGOJ:-KTkdWNH7C-f99oTCGOK:b-eodqo6) ([raw file](https://github.com/valloric/youcompleteme/blob/8263773ccde05a323443fa14f1c55d99ae88d26b/python/ycm/tests/diagnostic_filter_tests.py#L1)):*
> ```Python
> # Copyright (C) 2013  Google Inc.
> ```

Same as before, bad copyright line.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-:-KTkdafsqVmriqoMFFER:bxpnb11)*
<!-- Sent from Reviewable.io -->

dhleong commented:
Woops! Forgot I was running the tests with the `--no-flake8` flag (had some trouble getting the test environment working at first). 

I've fixed the copyright and lint issues, and went ahead and took a stab at some documentation as well.

dhleong commented:
Aaaand I see the travis errors. Looking into them now

dhleong commented:
Fixed Travis.

On Mon, Oct 10, 2016, 6:14 PM Val Markovic <notifications@github.com> wrote:

> Thanks for the PR! :)
>
> I agree this is a feature we should have.
>
> Travis runs are failing for this PR, so you'll need to look into that.
>
> I gave a light review; the code doesn't quite match the style we follow in
> the rest of the codebase. I suggest looking through other files to identify
> what the style looks like. That might be faster than us pointing out every
> style violation. :)
>
> Also, this needs to change the README so that we have documentation for
> the new option.
> ------------------------------
>
> Review status: 0 of 4 files reviewed at latest revision, 4 unresolved
> discussions, some commit checks failed.
> ------------------------------
>
> *python/ycm/diagnostic_filter.py, line 1 at r1
> <https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTkd6g70Aoq1XQS4g3s:-KTkd6g8tS1lCzJCXZCc:b-boinl4>
> (raw file
> <https://github.com/valloric/youcompleteme/blob/8263773ccde05a323443fa14f1c55d99ae88d26b/python/ycm/diagnostic_filter.py#L1>):*
>
> # Copyright (C) 2013  Google Inc.
>
> The copyright line is wrong. Update year and author is "YouCompleteMe
> contributors".
> ------------------------------
>
> *python/ycm/diagnostic_filter.py, line 41 at r1
> <https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTkdNyA0frVaC5k1iCd:-KTkdNyA0frVaC5k1iCe:b69sgpp>
> (raw file
> <https://github.com/valloric/youcompleteme/blob/8263773ccde05a323443fa14f1c55d99ae88d26b/python/ycm/diagnostic_filter.py#L41>):*
>
>       actual_filter_type = filter_type
>
>       if filter_type[0] == '!':
>
> Whitespace inside brackets.
> ------------------------------
>
> *python/ycm/diagnostic_filter.py, line 45 at r1
> <https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTkdEqYZBHPtqGNPCAI:-KTkdEqYZBHPtqGNPCAJ:b-gak7ss>
> (raw file
> <https://github.com/valloric/youcompleteme/blob/8263773ccde05a323443fa14f1c55d99ae88d26b/python/ycm/diagnostic_filter.py#L45>):*
>
>         filter_type = filter_type[1:]
>       compiler = FILTER_COMPILERS.get( filter_type )
>
> No trailing whitespace please. :)
> ------------------------------
>
> *python/ycm/tests/diagnostic_filter_tests.py, line 1 at r1
> <https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTkdWNH7C-f99oTCGOJ:-KTkdWNH7C-f99oTCGOK:b-eodqo6>
> (raw file
> <https://github.com/valloric/youcompleteme/blob/8263773ccde05a323443fa14f1c55d99ae88d26b/python/ycm/tests/diagnostic_filter_tests.py#L1>):*
>
> # Copyright (C) 2013  Google Inc.
>
> Same as before, bad copyright line.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-:-KTkdafsqVmriqoMFFER:bxpnb11>*
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Valloric/YouCompleteMe/pull/2377#issuecomment-252760257>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAx0FricvR59rZ5-9y9Hq01ZxWcD-27Nks5qyrjCgaJpZM4KS3Tn>
> .
>

Valloric commented:
Some more comments; looking good so far, please continue. :)

---

Review status: 0 of 5 files reviewed at latest revision, 6 unresolved discussions.

---

*[doc/youcompleteme.txt, line 2069 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTqB7NyWSL-HB-U1wPx:-KTqB7NyWSL-HB-U1wPy:b-q72t37) ([raw file](https://github.com/valloric/youcompleteme/blob/76784b2a6ce61a3c0ebc8735fdf8592bbb0d1767/doc/youcompleteme.txt#L2069)):*
> ```
> <
> -------------------------------------------------------------------------------
> The *g:ycm_quiet_messages* option
> ```

I really hope you didn't edit this file by hand because we only edit the README.md file; this one is auto-generated with a tool. :)

So don't worry about this file, just add your docs to the README file.

---

*[doc/youcompleteme.txt, line 2076 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTqBVmu91psJ3YKd63H:-KTqBVmu91psJ3YKd63I:b-g0e8ze) ([raw file](https://github.com/valloric/youcompleteme/blob/76784b2a6ce61a3c0ebc8735fdf8592bbb0d1767/doc/youcompleteme.txt#L2076)):*
> ```
> >
>   let g:ycm_quiet_messages = {
>         \ "!level":  "errors",
> ```

Why is there what looks like a pipe character in "level"?

---

*[python/ycm/diagnostic_filter.py, line 49 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTqC1HOxAPAdIOy5npm:-KTqC1HOxAPAdIOy5npn:b-khhu81) ([raw file](https://github.com/valloric/youcompleteme/blob/76784b2a6ce61a3c0ebc8735fdf8592bbb0d1767/python/ycm/diagnostic_filter.py#L49)):*
> ```Python
> 
> 
>   def Accept( self, diagnostic ):
> ```

Should probably be called `IsAllowed` to stress that this returns a bool.

---

*[python/ycm/diagnostic_filter.py, line 50 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTqCeZiHxzKRyH6g-EJ:-KTqCeZiHxzKRyH6g-EK:b-n54n0i) ([raw file](https://github.com/valloric/youcompleteme/blob/76784b2a6ce61a3c0ebc8735fdf8592bbb0d1767/python/ycm/diagnostic_filter.py#L50)):*
> ```Python
> 
>   def Accept( self, diagnostic ):
>     # NB: we Accept() the diagnostic ONLY if
> ```

I'd suggest using "Note:" instead of "NB:"; not everyone knows that it stands for Nota Bene or what that even means. :)

---

*[python/ycm/diagnostic_filter.py, line 108 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTqCP8zBVlZRt5J8uky:-KTqCP9-OZnY4qabTrtl:b8je9eh) ([raw file](https://github.com/valloric/youcompleteme/blob/76784b2a6ce61a3c0ebc8735fdf8592bbb0d1767/python/ycm/diagnostic_filter.py#L108)):*
> ```Python
> 
> 
> FILTER_COMPILERS  = { 'regex'      : _CompileRegex,
> ```

Why the extra whitespace before `:`? Does not seem useful...

---

*[python/ycm/tests/diagnostic_filter_tests.py, line 40 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTqBqwJChnte9BlQ7qH:-KTqBqwKip8zRW8NLHmO:b31s7c5) ([raw file](https://github.com/valloric/youcompleteme/blob/76784b2a6ce61a3c0ebc8735fdf8592bbb0d1767/python/ycm/tests/diagnostic_filter_tests.py#L40)):*
> ```Python
> 
> 
> def _assert_accepts( filter, text ):
> ```

Should probably be _CamelCase. Similar for the next function.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-:-KTqCxYchkj_grbP08Nk:b-iq2s0i)*
<!-- Sent from Reviewable.io -->

micbou commented:
I find the name `ycm_quiet_messages` confusing. Which messages? Does it silence all messages? A better name would be `ycm_filter_diagnostics`.

I think that trying to be compatible with Syntastic is a mistake as it limits our options. For instance, instead of defining the `g:ycm_*_quiet_messages` options, we could make the `ycm_filter_diagnostics` option a dictionary with filetypes as keys (like [the `g:ycm_semantic_triggers` option](https://github.com/Valloric/YouCompleteMe#the-gycm_semantic_triggers-option)). Also, the `regex` field uses Python regexes while Syntastic only supports Vim regexes, so I am not sure how a fallback to Syntastic is supposed to work.

---

Review status: 0 of 5 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-:-KTqDMVeLqdx9WGeeji6:b-f6g4rh)*
<!-- Sent from Reviewable.io -->

dhleong commented:
Review status: 0 of 5 files reviewed at latest revision, 10 unresolved discussions.

---

*[doc/youcompleteme.txt, line 2069 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTqB7NyWSL-HB-U1wPx:-KTqVM7I7Wh-ifOpkpgz:b-qj64bj) ([raw file](https://github.com/valloric/youcompleteme/blob/76784b2a6ce61a3c0ebc8735fdf8592bbb0d1767/doc/youcompleteme.txt#L2069)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> I really hope you didn't edit this file by hand because we only edit the README.md file; this one is auto-generated with a tool. :)
> 
> So don't worry about this file, just add your docs to the README file.
</details>
I did. Neat about the tool, though! I'll wait on updating this until we resolve @micbou's questions.

---

*[doc/youcompleteme.txt, line 2076 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTqBVmu91psJ3YKd63H:-KTqUrzEL5C9Bsn1lil4:bn6w38t) ([raw file](https://github.com/valloric/youcompleteme/blob/76784b2a6ce61a3c0ebc8735fdf8592bbb0d1767/doc/youcompleteme.txt#L2076)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Why is there what looks like a pipe character in "level"?
</details>
It's a bang to invert the rule ("silence things that are NOT errors")

---

*[python/ycm/diagnostic_filter.py, line 1 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTkd6g70Aoq1XQS4g3s:-KTqU7JgDqc3cczj-AOl:b-896fix) ([raw file](https://github.com/valloric/youcompleteme/blob/8263773ccde05a323443fa14f1c55d99ae88d26b/python/ycm/diagnostic_filter.py#L1)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> The copyright line is wrong. Update year and author is "YouCompleteMe contributors".
</details>
Done.

---

*[python/ycm/diagnostic_filter.py, line 41 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTkdNyA0frVaC5k1iCd:-KTqTzC8aNhnESyqhz4r:b-896fix) ([raw file](https://github.com/valloric/youcompleteme/blob/8263773ccde05a323443fa14f1c55d99ae88d26b/python/ycm/diagnostic_filter.py#L41)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Whitespace inside brackets.
</details>
Done.

---

*[python/ycm/diagnostic_filter.py, line 45 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTkdEqYZBHPtqGNPCAI:-KTqU5wqzVfYk3OXvG2E:b-896fix) ([raw file](https://github.com/valloric/youcompleteme/blob/8263773ccde05a323443fa14f1c55d99ae88d26b/python/ycm/diagnostic_filter.py#L45)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> No trailing whitespace please. :)
</details>
Done.

---

*[python/ycm/diagnostic_filter.py, line 108 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTqCP8zBVlZRt5J8uky:-KTqUU6w2Kj-kAYyim3X:b-xtbh1w) ([raw file](https://github.com/valloric/youcompleteme/blob/76784b2a6ce61a3c0ebc8735fdf8592bbb0d1767/python/ycm/diagnostic_filter.py#L108)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Why the extra whitespace before `:`? Does not seem useful...
</details>
Woops!

---

*[python/ycm/tests/diagnostic_filter_tests.py, line 1 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTkdWNH7C-f99oTCGOJ:-KTqUXJ7VEjI6dBiAWKj:b-896fix) ([raw file](https://github.com/valloric/youcompleteme/blob/8263773ccde05a323443fa14f1c55d99ae88d26b/python/ycm/tests/diagnostic_filter_tests.py#L1)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Same as before, bad copyright line.
</details>
Done.

---

*[python/ycm/tests/diagnostic_filter_tests.py, line 40 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTqBqwJChnte9BlQ7qH:-KTqUbCGSGGbuUSFPrOt:bu18l7h) ([raw file](https://github.com/valloric/youcompleteme/blob/76784b2a6ce61a3c0ebc8735fdf8592bbb0d1767/python/ycm/tests/diagnostic_filter_tests.py#L40)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Should probably be _CamelCase. Similar for the next function.
</details>
I was following hamcrest's style here, since it's an assertion, but I can change it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2377)*
<!-- Sent from Reviewable.io -->

dhleong commented:
@micbou All valid points. I actually expected it to be a map of filetype to (something) when I first looked into this, so I'm on board for ignoring the syntastic precedent here. 

If that's the case, should we completely drop the syntastic-style rule specification? Maybe something like:

```viml
let g:ycm_filter_diagnostics = {
  \ "java": [
  \      {
  \        "type": "regex",
  \        "regex": ".*taco.*"
  \      },
  \      {
  \         " more rules
  \      {
  \    ]
  \ }
```

I'm by no means attached to this (it's pretty verbose) just throwing out ideas

Valloric commented:
Since Syntastic uses vim regexes and we use Python regexes, we can't be compatible. So I'm fine with us doing our own thing.

---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[python/ycm/tests/diagnostic_filter_tests.py, line 40 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTqBqwJChnte9BlQ7qH:-KTtqTGIT002b_ozNfpU:b6dcv1w) ([raw file](https://github.com/valloric/youcompleteme/blob/76784b2a6ce61a3c0ebc8735fdf8592bbb0d1767/python/ycm/tests/diagnostic_filter_tests.py#L40)):*
<details><summary><i>Previously, dhleong (Daniel Leong) wrote…</i></summary>
> I was following hamcrest's style here, since it's an assertion, but I can change it.
</details>
Fine either way. One is consistent with our codebase, the other with hamcrest assertions.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-:-KTtqfVnooWDBBqNjTCk:b-26eoth)*
<!-- Sent from Reviewable.io -->

micbou commented:
@dhleong I am okay with that structure. The only other structure I can think of is this one:
```viml
let g:ycm_filter_diagnostics = {
  \ "java": {
  \      "regex": [ ".*taco.*", ... ],
  \      "level": "error",
  \      ...
  \    }
  \ }
```
but I am not sure that's better.

---

Reviewed 1 of 4 files at r2.
Review status: 1 of 5 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-:-KU2pGi3fPJkT7GmprN9:b-aeu3dy)*
<!-- Sent from Reviewable.io -->

dhleong commented:
Sorry I'm being so slow, not at home for the next week-ish.

I'm still trying to brainstorm the clearest, yet still succinct, syntax.

What about:

```viml
let g:ycm_filter_diagnostics = {
  \ "java": {
  \   "accept": {
  \     "level": "error"
  \   },
  \   "reject": {
  \     "regex": [ ".*taco.*" ]
  \    }
  \ }
```

ALL `accept` rules specified MUST be matched. NO `reject` rules may match. `accept-any` is another possibility which, as you might expect, would only allow items for which *at least one* rule matched.

We could get more verbose with the keys: `accept` could be `must-match-all`; `reject` could be `must-not-match-any,` etc. This could be a bit annoying to type out (though we could also just use spaces to improve this, eg `must match all`) but might be more obvious to maintain.

Either way, I think the first idea I posted (with the `type` field) would be annoying to maintain in a vimrc, even if it'd be simpler to implement. I also think syntastic's `!` syntax is not as intuitive as the `reject`/`accept` proposed above.

Valloric commented:
@dhleong Instead of doing accept/reject, I think reject is enough. So @micbou's proposal would work well. I don't see the benefit of listing accept filters; everything starts off being accepted, and filtering is there to remove things that match. So if any filter matches, we remove the diagnostic.

dhleong commented:
The idea of "accept" is to set a baseline, I think; it replaces the "!"
(invert) operator from Syntastic. It means "reject everything that DOES NOT
match this," which is why I think the more verbose "must match
all/each/every" key may more clearly indicate how it can be used. You can
then filter out specific diagnostics with the "reject" key.

I can't think if a really great use for this at the moment (other than only
accepting errors, but that's just a semantically different way of saying
"reject warnings," since we only have those two types right now, as far as
I know), but I imagine Syntastic wouldn't have implemented it for no
reason....

On Fri, Oct 14, 2016, 3:14 PM Val Markovic <notifications@github.com> wrote:

> @dhleong <https://github.com/dhleong> Instead of doing accept/reject, I
> think reject is enough. So @micbou <https://github.com/micbou>'s proposal
> would work well. I don't see the benefit of listing accept filters;
> everything starts off being accepted, and filtering is there to remove
> things that match. So if any filter matches, we remove the diagnostic.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Valloric/YouCompleteMe/pull/2377#issuecomment-253953152>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAx0FlE6Fh1oR4jcLiD8dAvzaVNj8C1wks5q0CkQgaJpZM4KS3Tn>
> .
>

Valloric commented:
> I can't think if a really great use for this at the moment

Neither can I and that's a pretty good reason to not add it right now. [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) :).

>  but I imagine Syntastic wouldn't have implemented it for no
reason....

I'm not going to assume why they added it, all I can see is that there doesn't seem to be a compelling use-case for it in YCM.

dhleong commented:
> YAGNI :).

Fair enough. I've refactored the code to use @micbou 's recommendation, and updated README.md

micbou commented:
I tried the changes myself and it works great. Only issue is that the `:YcmDiags` command still adds all diagnostics in the location list and I think a user would expect them to be filtered too.

---

Reviewed 1 of 3 files at r4, 3 of 4 files at r5.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[python/ycm/diagnostic_filter.py, line 50 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTqCeZiHxzKRyH6g-EJ:-KUI5W-UYs0LjNcHlZpW:bnd888d) ([raw file](https://github.com/valloric/youcompleteme/blob/76784b2a6ce61a3c0ebc8735fdf8592bbb0d1767/python/ycm/diagnostic_filter.py#L50)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> I'd suggest using "Note:" instead of "NB:"; not everyone knows that it stands for Nota Bene or what that even means. :)
</details>
Another reason to use `NOTE:` is that it is highlighted in Vim. `NB:` is used in other places (there is still one in this file and three in the tests). Could you replace them?

---

*[python/ycm/diagnostic_filter.py, line 35 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUI4WMTDWP6L6KbqE7z:-KUI4WMTDWP6L6KbqE8-:bf91mbd) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/diagnostic_filter.py#L35)):*
> ```Python
> 
>     for filter_type in iterkeys( config ):
>       actual_filter_type = filter_type
> ```

Is this temporary variable really needed?

---

*[python/ycm/diagnostic_filter.py, line 40 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUHywDREbMmRp-KU6zR:-KUHywDSKDSFVP6fw53w:b-nk49lu) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/diagnostic_filter.py#L40)):*
> ```Python
>       if compiler is not None:
>         for filter_config in _ListOf( config[ actual_filter_type ] ):
>           fn = compiler( filter_config )
> ```

A more explicit variable name than `fn` should be used. 

---

*[python/ycm/diagnostic_filter.py, line 46 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUHyYZrvZkjAzS3udad:-KUHyYZsoYGw6SQhJRzH:by3jfkm) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/diagnostic_filter.py#L46)):*
> ```Python
>   def IsAllowed( self, diagnostic ):
>     # NOTE: a diagnostic IsAllowed() ONLY if
>     #  NO filters match it
> ```

There is enough room to put this comment on one line.

---

*[python/ycm/diagnostic_filter.py, line 47 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUDv9q1AFkXGWwtaT8X:-KUDv9q2s2JpJXMq2oCt:bk0s6y8) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/diagnostic_filter.py#L47)):*
> ```Python
>     # NOTE: a diagnostic IsAllowed() ONLY if
>     #  NO filters match it
>     for f in self._filters:
> ```

`f` is not explicit enough as a variable name.

---

*[python/ycm/diagnostic_filter.py, line 55 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUDsOYSDA2PyIMnenkE:-KUDsOYSDA2PyIMnenkF:bue5lfu) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/diagnostic_filter.py#L55)):*
> ```Python
> 
>   @staticmethod
>   def from_filetype( user_options, filetypes ):
> ```

We use CamelCase for function (and method) names but lower_case_with_underscores for variable names. This applies for the `typeSpec` and `filterValue` variables defined in this method.

---

*[python/ycm/diagnostic_filter.py, line 60 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUDvRHTlQfX_RtP3AoW:-KUDvRHTlQfX_RtP3AoX:bdoqldv) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/diagnostic_filter.py#L60)):*
> ```Python
>     for typeSpec, filterValue in iteritems( dict( all_filters ) ):
>       if typeSpec.find(',') != -1:
>         for ft in typeSpec.split(','):
> ```

`filetype` is more explicit than `ft`.

---

*[python/ycm/diagnostic_filter.py, line 80 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUI6y24EBRv5TzzHHf6:-KUI6y24EBRv5TzzHHf7:b4chf5) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/diagnostic_filter.py#L80)):*
> ```Python
> 
> def _Merge( into, other ):
>   for k in iterkeys( other ):
> ```

`key` instead of `k`?

---

*[python/ycm/diagnostic_filter.py, line 81 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUDwN372_jXvH7hVocl:-KUDwN372_jXvH7hVocm:b-ilnluy) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/diagnostic_filter.py#L81)):*
> ```Python
> def _Merge( into, other ):
>   for k in iterkeys( other ):
>     into[k] = _ListOf( into.get( k ) ) + _ListOf( other[ k ] )
> ```

Spaces inside square brackets.

---

*[python/ycm/diagnostic_interface.py, line 60 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUHtpQJC4philn0RhWQ:-KUHtpQKlUvEjLDoDmxS:b5hzrlh) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/diagnostic_interface.py#L60)):*
> ```Python
> 
>   def GetWarningCount( self ):
>     return len( self._FilterDiagnostics( _DiagnosticIsWarning ) )
> ```

Is there a way to use the new `DiagnosticFilter` class in these two methods instead of the `_FilterDiagnostics`, `_DiagnosticIsError`, and `_DiagnosticIsWarning` functions?

---

*[python/ycm/diagnostic_interface.py, line 71 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUHslPTrTafs72Urnh3:-KUHslPUZ63gStK3GYJu:b-h18ido) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/diagnostic_interface.py#L71)):*
> ```Python
>     diag_filter = DiagnosticFilter.from_filetype(
>             self._user_options,
>             vimsupport.CurrentFiletypes() )
> ```

Creating this object each time there are new diagnostics is not really efficient. It could be cached in a dictionary with filetypes as keys.

---

*[python/ycm/tests/diagnostic_filter_tests.py, line 1 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUI0ehDZV81Z6xMQRPk:-KUI0ehEJqJwh8bBqfLT:blkv3ca) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/tests/diagnostic_filter_tests.py#L1)):*
> ```Python
> # Copyright (C) 2016  YouCompleteMe contributors
> ```

Could you rename this file to `diagnostic_filter_test.py` (without a `s` to `test`)? Just to be consistent with the other files in the tests folder.

---

*[python/ycm/tests/diagnostic_filter_tests.py, line 27 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUI-u4wL78zPhIyUsa2:-KUI-u4xYNw9qjO2Xwq0:b-7lxd3t) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/tests/diagnostic_filter_tests.py#L27)):*
> ```Python
> 
> from ycm.test_utils import MockVimModule
> MockVimModule()
> ```

I think these two lines can be removed as there is no direct interaction with Vim in the tests. This should fix the builds by the way.

---

*[python/ycm/tests/diagnostic_filter_tests.py, line 60 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUDqphu6rnik8roLfWd:-KUDqphvENm0QfQg_NU5:b-2aj3rq) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/tests/diagnostic_filter_tests.py#L60)):*
> ```Python
> 
> 
> class ListOrSingle_test():
> ```

Is this really needed to put these tests inside a class? I don't see the benefit. Same comment for the `Level_test` class below. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-:-KUDxDbng7-Lwmq9nvrl:b-ubyj6z)*
<!-- Sent from Reviewable.io -->

dhleong commented:
Got it! Will look into `:YcmDiags`

---

Review status: 3 of 5 files reviewed at latest revision, 15 unresolved discussions.

---

*[python/ycm/diagnostic_interface.py, line 60 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUHtpQJC4philn0RhWQ:-KUhIkbcbB6lFeMskAVz:b-romcp9) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/diagnostic_interface.py#L60)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> Is there a way to use the new `DiagnosticFilter` class in these two methods instead of the `_FilterDiagnostics`, `_DiagnosticIsError`, and `_DiagnosticIsWarning` functions?
</details>
Yes, I believe so. I'll look into this next.

---

*[python/ycm/diagnostic_interface.py, line 71 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUHslPTrTafs72Urnh3:-KUh7oMbqMyAvhbwNgaf:b-uib5ex) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/diagnostic_interface.py#L71)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> Creating this object each time there are new diagnostics is not really efficient. It could be cached in a dictionary with filetypes as keys.
</details>
Agreed, but I was unsure on how to best handle changing config values and filetypes—it seems like the `YouCompleteMe` object is a singleton? I guess I can assume the configuration doesn't change between calls, but the filetype still could, which means the `DiagnosticFilter` class would no longer be able to be nice and stateless.

Perhaps we could compile all the filters for each type up front, then build the specific list of filters (wrapped in a `DiagnosticFilter` or similar class), then assemble (with cache) the full list for all current filetypes at the call site? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-:-KUhIxnFXkaWx6eXYWjK:b28dfp3)*
<!-- Sent from Reviewable.io -->

dhleong commented:
Addressed the `:YcmDiags` issue with a `_ApplyDiagnosticFilter` method that filters a provided `diags` list using the current filetypes, shared with the other usage. I think it may be able to be tweaked for use with `_FilterDiagnostics` as well; the `_DiagnosticIsError` and `_DiagnosticIsWarning` vars can be initialized by making the filter compiler functions public and use `_CompileLevel`.

---

Review status: 2 of 5 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[doc/youcompleteme.txt, line 2069 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTqB7NyWSL-HB-U1wPx:-KUi_-MI1Y_xfn2OhYYB:bavi39a) ([raw file](https://github.com/valloric/youcompleteme/blob/76784b2a6ce61a3c0ebc8735fdf8592bbb0d1767/doc/youcompleteme.txt#L2069)):*
<details><summary><i>Previously, dhleong (Daniel Leong) wrote…</i></summary>
> I did. Neat about the tool, though! I'll wait on updating this until we resolve @micbou's questions.
</details>
Done. (see the changes to README)

---

*[python/ycm/diagnostic_filter.py, line 50 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KTqCeZiHxzKRyH6g-EJ:-KUiZpgI5blDGXHuiIQj:b-896fix) ([raw file](https://github.com/valloric/youcompleteme/blob/76784b2a6ce61a3c0ebc8735fdf8592bbb0d1767/python/ycm/diagnostic_filter.py#L50)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> Another reason to use `NOTE:` is that it is highlighted in Vim. `NB:` is used in other places (there is still one in this file and three in the tests). Could you replace them?
</details>
Done.

---

*[python/ycm/diagnostic_filter.py, line 35 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUI4WMTDWP6L6KbqE7z:-KUiZwrAMY11uxzBB6t4:b-896fix) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/diagnostic_filter.py#L35)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> Is this temporary variable really needed?
</details>
Done.

---

*[python/ycm/diagnostic_filter.py, line 40 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUHywDREbMmRp-KU6zR:-KUiZulzJ1cq92SenLLw:b-896fix) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/diagnostic_filter.py#L40)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> A more explicit variable name than `fn` should be used. 
</details>
Done.

---

*[python/ycm/diagnostic_filter.py, line 46 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUHyYZrvZkjAzS3udad:-KUiZo6Kn1niMvqg_VYV:b-896fix) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/diagnostic_filter.py#L46)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> There is enough room to put this comment on one line.
</details>
Done.

---

*[python/ycm/diagnostic_filter.py, line 47 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUDv9q1AFkXGWwtaT8X:-KUiZaBj8MZPVFf3yGS1:b-adt9xz) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/diagnostic_filter.py#L47)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> `f` is not explicit enough as a variable name.
</details>
Done. I used camelCase because it's a local variable reference to a function, but I can change it to snakey_case if you still prefer that in this case.

---

*[python/ycm/diagnostic_filter.py, line 55 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUDsOYSDA2PyIMnenkE:-KUiZZ8_uWAbJCAHcAGj:b-896fix) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/diagnostic_filter.py#L55)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> We use CamelCase for function (and method) names but lower_case_with_underscores for variable names. This applies for the `typeSpec` and `filterValue` variables defined in this method.
</details>
Done.

---

*[python/ycm/diagnostic_filter.py, line 60 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUDvRHTlQfX_RtP3AoW:-KUiZ_5LVDXsSfjUbMgH:b-896fix) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/diagnostic_filter.py#L60)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> `filetype` is more explicit than `ft`.
</details>
Done.

---

*[python/ycm/diagnostic_filter.py, line 80 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUI6y24EBRv5TzzHHf6:-KUiZVNZMvTwx7kcWxJj:bypsnhe) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/diagnostic_filter.py#L80)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> `key` instead of `k`?
</details>
Done. (Removed the whole thing)

---

*[python/ycm/diagnostic_filter.py, line 81 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUDwN372_jXvH7hVocl:-KUiZUpuFvKBzLEJKKSE:b-896fix) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/diagnostic_filter.py#L81)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> Spaces inside square brackets.
</details>
Done.

---

*[python/ycm/diagnostic_interface.py, line 71 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUHslPTrTafs72Urnh3:-KUiYutfmapzacvvJueh:b-p3003v) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/diagnostic_interface.py#L71)):*
<details><summary><i>Previously, dhleong (Daniel Leong) wrote…</i></summary>
> Agreed, but I was unsure on how to best handle changing config values and filetypes—it seems like the `YouCompleteMe` object is a singleton? I guess I can assume the configuration doesn't change between calls, but the filetype still could, which means the `DiagnosticFilter` class would no longer be able to be nice and stateless.
> 
> Perhaps we could compile all the filters for each type up front, then build the specific list of filters (wrapped in a `DiagnosticFilter` or similar class), then assemble (with cache) the full list for all current filetypes at the call site? 
</details>
Pushed a change that implements my proposed idea and resolves (I think) most of the other notes.

---

*[python/ycm/tests/diagnostic_filter_test.py, line 1 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUI0ehDZV81Z6xMQRPk:-KUi_mf_qtlKd5PUN3ds:b-896fix) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/tests/diagnostic_filter_test.py#L1)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> Could you rename this file to `diagnostic_filter_test.py` (without a `s` to `test`)? Just to be consistent with the other files in the tests folder.
</details>
Done.

---

*[python/ycm/tests/diagnostic_filter_test.py, line 27 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUI-u4wL78zPhIyUsa2:-KUi_on9ufu1YL--WAMF:b-ipu508) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/tests/diagnostic_filter_test.py#L27)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> I think these two lines can be removed as there is no direct interaction with Vim in the tests. This should fix the builds by the way.
</details>
Needed now, given `_MasterDiagnosticFilter`'s implementation of `IsAccepted`; the build works for me locally—still trying to figure out why the Travis build failed importing `ycm.test_utils` just now, when the adjacent tests imported it just fine.

---

*[python/ycm/tests/diagnostic_filter_test.py, line 60 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUDqphu6rnik8roLfWd:-KUiZHh7564x-rfMD1-w:b-gwca0) ([raw file](https://github.com/valloric/youcompleteme/blob/df388c77f4c091d78357d157df7c6cb748128518/python/ycm/tests/diagnostic_filter_test.py#L60)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> Is this really needed to put these tests inside a class? I don't see the benefit. Same comment for the `Level_test` class below. 
</details>
Just for organization. No strong need; removed classes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-:-KUiY-Za7IG-a1RQR9ad:bcd6ddp)*
<!-- Sent from Reviewable.io -->

dhleong commented:
Just rebased against master and pushed an update that should fix the tests. I guess the test_utils module moved on master which broke the tests on travis but not locally for me.

micbou commented:
It's almost ready for me. One remaining thing to do is to use [the `html2vimdoc.py` script to update the Vim documentation](https://github.com/xolox/vim-tools#html2vimdoc) but we can take care of that if you want. In this case, could you revert the changes done to `doc/youcompleteme.txt`?

---

Reviewed 2 of 2 files at r6, 1 of 3 files at r7, 3 of 3 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[python/ycm/diagnostic_filter.py, line 39 at r8](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUktgob4vg7ZEsdRYa4:-KUktgocBpMb0T4345Ys:b-ah2bxa) ([raw file](https://github.com/valloric/youcompleteme/blob/e2852a8b2b8419acc453950e9d60e9187dd7d54c/python/ycm/diagnostic_filter.py#L39)):*
> ```Python
>     else:
>       self._filters = _CompileFilters( config_or_filters )
>       print( 'CompileFilters', config_or_filters)
> ```

These two prints should be removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-:-KUlFoBqMb3jZW1rDbc7:b-43hjps)*
<!-- Sent from Reviewable.io -->

dhleong commented:
No worries, I went ahead and just updated the .txt using the script. I also updated the `GetErrorCount` etc to use the new filtering stuff.

---

Review status: 2 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[python/ycm/diagnostic_filter.py, line 39 at r8](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUktgob4vg7ZEsdRYa4:-KUm0ksyT627oXPYXd-9:bs7wumx) ([raw file](https://github.com/valloric/youcompleteme/blob/e2852a8b2b8419acc453950e9d60e9187dd7d54c/python/ycm/diagnostic_filter.py#L39)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> These two prints should be removed.
</details>
Done. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-:-KUm0XFEL8-JYm4hmyaY:b4do7r4)*
<!-- Sent from Reviewable.io -->

micbou commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r9.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-:-KUm5ezBPxOKvzfQCc45:bnfp4nl)*
<!-- Sent from Reviewable.io -->

Valloric commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with minor comments. Once those are resolved, we can merge this.

Thanks for working on this!

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[README.md, line 1809 at r9](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUmHyr6aQNp_IfG8WmG:-KUmHyr6aQNp_IfG8WmH:bfay2uh) ([raw file](https://github.com/valloric/youcompleteme/blob/04c48e48c40cce3b4261cafee6c51a126ef5f354/README.md#L1809)):*
> ```Markdown
> 
> This option controls which diagnostics will be rendered by YCM. This option holds
> a dictionary of key-values, where the keys are Vim's filetype strings delimited by
> ```

Is this wrapped at 80 columns?

---

*[README.md, line 1815 at r9](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUmIMa4ZVcRs47MCgdP:-KUmIMa5cT_w-cqaY1ns:b-6yvoez) ([raw file](https://github.com/valloric/youcompleteme/blob/04c48e48c40cce3b4261cafee6c51a126ef5f354/README.md#L1815)):*
> ```Markdown
> and the value is a list of arguments to that filter. In the case of just a single 
> item in the list, you may omit the brackets and just provide the argument directly.
> If any filter matches a diagnostic, it will be dropped and YCM will not render it.
> ```

Maybe "display" instead of "render"?

---

*[python/ycm/diagnostic_filter.py, line 126 at r9](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUmIqo99A1UPu7HmWRX:-KUmIqoAYApOIzNzFw0w:b-gu7vzk) ([raw file](https://github.com/valloric/youcompleteme/blob/04c48e48c40cce3b4261cafee6c51a126ef5f354/python/ycm/diagnostic_filter.py#L126)):*
> ```Python
>   # valid kinds are WARNING and ERROR;
>   #  expected input levels are `warning` and `error`
>   # NOTE: we don't validate the input...
> ```

Why don't we? It shouldn't be hard to check that the string is either `WARNING` or `ERROR` after `.upper()`. If it's not one of the two, throw a ValueError exception.

---

*[python/ycm/tests/diagnostic_filter_test.py, line 69 at r9](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUmJJt4gaUH3vDTUBSv:-KUmJJt4gaUH3vDTUBSw:b-83m4ml) ([raw file](https://github.com/valloric/youcompleteme/blob/04c48e48c40cce3b4261cafee6c51a126ef5f354/python/ycm/tests/diagnostic_filter_test.py#L69)):*
> ```Python
> 
>   _assert_rejects( f, 'This is a Taco' )
>   _assert_accepts( f, 'This is a Burrito' )
> ```

I love the test data. :D

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-:-KUmJXgUHAOB85Lh7LkE:bxrlhl7)*
<!-- Sent from Reviewable.io -->

micbou commented:
Review status: 4 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[python/ycm/diagnostic_filter.py, line 126 at r9](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUmIqo99A1UPu7HmWRX:-KUmK8RBV-pB58_5yT_E:b-8bdxe0) ([raw file](https://github.com/valloric/youcompleteme/blob/04c48e48c40cce3b4261cafee6c51a126ef5f354/python/ycm/diagnostic_filter.py#L126)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Why don't we? It shouldn't be hard to check that the string is either `WARNING` or `ERROR` after `.upper()`. If it's not one of the two, throw a ValueError exception.
</details>
We should avoid throwing an exception as this will result in a traceback in Vim if we don't catch it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2377)*
<!-- Sent from Reviewable.io -->

dhleong commented:
Review status: 4 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[README.md, line 1809 at r9](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUmHyr6aQNp_IfG8WmG:-KUmQ1GNAY_mmml6wmuQ:bt2vr1b) ([raw file](https://github.com/valloric/youcompleteme/blob/04c48e48c40cce3b4261cafee6c51a126ef5f354/README.md#L1809)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Is this wrapped at 80 columns?
</details>
Fixed.

---

*[README.md, line 1815 at r9](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUmIMa4ZVcRs47MCgdP:-KUmQ26sbIQBkE1dgK1o:b-896fix) ([raw file](https://github.com/valloric/youcompleteme/blob/04c48e48c40cce3b4261cafee6c51a126ef5f354/README.md#L1815)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Maybe "display" instead of "render"?
</details>
Done.

---

*[python/ycm/diagnostic_filter.py, line 126 at r9](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUmIqo99A1UPu7HmWRX:-KUmQ3aZRglY-bMGkkV6:b-uytwh1) ([raw file](https://github.com/valloric/youcompleteme/blob/04c48e48c40cce3b4261cafee6c51a126ef5f354/python/ycm/diagnostic_filter.py#L126)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> We should avoid throwing an exception as this will result in a traceback in Vim if we don't catch it.
</details>
@micbou's concern is why I didn't validate, but I've looked at it again, and the exceptions *are* caught, since it's part of the YouCompleteMe init. It's a bit of a bummer to stop the world if the filter is invalid, but perhaps not *terrible*....

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2377)*
<!-- Sent from Reviewable.io -->

dhleong commented:
Review status: 4 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[python/ycm/diagnostic_filter.py, line 126 at r9](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-KUmIqo99A1UPu7HmWRX:-KUmU4SagBKr1yjGaTRr:b-l73xi5) ([raw file](https://github.com/valloric/youcompleteme/blob/04c48e48c40cce3b4261cafee6c51a126ef5f354/python/ycm/diagnostic_filter.py#L126)):*
<details><summary><i>Previously, dhleong (Daniel Leong) wrote…</i></summary>
> @micbou's concern is why I didn't validate, but I've looked at it again, and the exceptions *are* caught, since it's part of the YouCompleteMe init. It's a bit of a bummer to stop the world if the filter is invalid, but perhaps not *terrible*....
</details>
Here's a quick diff that adds some nice validation, indicating where in the map the error is (if we do want to go ahead and raise exceptions):

https://gist.github.com/dhleong/af3badd903151c8859d82ed1a51f4902

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2377)*
<!-- Sent from Reviewable.io -->

Valloric commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I'm fine with this without the validation. 

Thanks for working on this! 

@homu r=micbou

---

Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2377#-:-KUoMG94x5A81xIizs5E:b6rhgm4)*
<!-- Sent from Reviewable.io -->

homu commented:
:pushpin: Commit 7e12c12 has been approved by `micbou`

<!-- @homu r=micbou 7e12c127cf8ef3e10349b89a9aefd682869c4dbd -->

homu commented:
:hourglass: Testing commit 7e12c12 with merge f27787f...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/Valloric/YouCompleteMe/builds/170012501)
